packages/core/lib/console-child.js

```diff
@@ -33,7 +33,11 @@ function main() {
   });
 
   runCommand(command, options)
-    .then(() => process.exit(0))
+    .then(returnStatus => {
+      process.exitCode = returnStatus;
+      return require("@truffle/promise-tracker").waitForOutstandingPromises();
```

Comment:
It's called with no params, so, isn't `iterable` in the implementation is just an empty array. Unless, I'm missing some setup in this scenario :/

```
export async function waitForOutstandingPromises<T extends Object>(
  options?: WaitForOutstandingPromiseOptions<T>
) {
  const { target } = options || {};
  let { catchHandler } = options || {};

  const iterable =
    (target
      ? _outstandingPromiseInstanceMap.get(target)?.values()
      : _outstandingPromiseMap.keys()) || [];

  catchHandler = catchHandler || (() => {});

  await Promise.all(Array.from(iterable, p => p.catch(catchHandler)));
}
```

Reply:
Yeah, hrm. Something weird must be implicitly happening, because this works and matches the usage in cli.js that @benjamincburns put there.

Ben, can you teach us?

Reply:
tl;dr: This is being used as expected here.

> It’s called with no params, so, isn’t iterable in the implementation just an empty array?

When `waitForOutstandingPromises` is called with no arguments (as it should be in all non-test use cases), the value of `iterable` is only an empty array when there are no outstanding promises on which to wait.

The object destructuring assignment for `target` coupled with the chained ternary operators probably make this horrendously unreadable. Sorry about that. Allow me to break things down a bit.

First off, on the line where we initialize `target`:

```typescript
const { target } = options || {};
```

We wind up with `target` being assigned `undefined`. This happens because `options` is `undefined`, so we fallback to the empty object literal and then attempt to destructure the `target` property out of that empty object literal. The empty object literal has no `target` property, so the destructuring assignment results with `target` being initialized with a value of `undefined`.

Given that detail, the following statement should be a bit more clear:

```typescript
  const iterable =
    (target
      ? _outstandingPromiseInstanceMap.get(target)?.values()
      : _outstandingPromiseMap.keys()) || [];
```

The value of `target` is `undefined`, so we branch to the "else" statement in the ternary operator, which assigns the return value of `_outstandingPromiseMap.keys()` to the `iterable` variable. This always assigns that return value, as `_outstandingPromiseMap` is a module-scoped variable, and it's initialized as `new Map<Promise<any>, boolean>()` on module load.

If there's any further uncertainty, be sure to check [the tests for the `promise-tracker` package](https://github.com/trufflesuite/truffle/blob/40502a245a2a174e519e461cfc7478b08d1df7bd/packages/promise-tracker/test/tracker.test.ts) as they're fairly comprehensive and they demonstrate and guard its usage nicely.

@gnidan @cds-amal @cliffoo please let me know if I can be of further assistance.

Reply:
Thanks @benjamincburns, my question relates to how `_outstandingPromiseMap` is seeded with values. If nothing is inserted, then how does it know what to track? This PR only calls the tracking code (as a JS module) and doesn't have any code to set/identify/decorate what has to be awaited.

Reply:
There's a [README.md](https://github.com/trufflesuite/truffle/blob/ec3ba9c0417dad7f1f323085b7c930edc15e7427/packages/promise-tracker/README.md), if that helps?

Otherwise the necessary code in question is already decorated. `dashboard-message-bus-client` depends on `promise-tracker`, and its internal `send` method is tracked [here](https://github.com/trufflesuite/truffle/blob/b371b9d4e33ae12e2fd1dd9a1c8990ca7570c207/packages/dashboard-message-bus-client/lib/connection/index.ts#L216-L231).

Decorating it with the `@tracked` label is all that's necessary to cause it to be tracked.

Reply:
Also if it helps any, [this is the function that does the actual task tracking](https://github.com/trufflesuite/truffle/blob/b371b9d4e33ae12e2fd1dd9a1c8990ca7570c207/packages/promise-tracker/lib/index.ts#L17-L31).

The `tracked` decorator injects this logic by [returning a property descriptor](https://github.com/trufflesuite/truffle/blob/b371b9d4e33ae12e2fd1dd9a1c8990ca7570c207/packages/promise-tracker/lib/index.ts#L33-L36) that replaces [the decorated method implementation](https://github.com/trufflesuite/truffle/blob/b371b9d4e33ae12e2fd1dd9a1c8990ca7570c207/packages/promise-tracker/lib/index.ts#L13) with a closure over that impl that proxies calls to it and handles tracking of outstanding promises by wrapping any returned promises in special promise handlers that manage the state of `_outstandingPromiseMap` and friends.

Reply:
> 
> Otherwise the necessary code in question is already decorated. `dashboard-message-bus-client` depends on `promise-tracker`, and its internal `send` method is tracked [here](https://github.com/trufflesuite/truffle/blob/b371b9d4e33ae12e2fd1dd9a1c8990ca7570c207/packages/dashboard-message-bus-client/lib/connection/index.ts#L216-L231).
> 
Thanks @benjamincburns, this is the piece I was missing. It was unclear from the code that it relied on an already coded setup.